### PR TITLE
Fix for list pin in integrated circuits

### PR DESCRIPTION
**Pull request by arsserpentarium**
I am very sorry for problems with list operations.Now it's fixed.
Also fixed problem with external battery sensor.
:cl: 
fix: now lists should work properly
/:cl:

### CODE REVIEW

File: code/modules/integrated_electronics/core/pins.dm [hunk 1]
```diff
@@ -115,7 +115,7 @@ list[](
 		holder.on_data_written()
 	else if(islist(new_data))
 		var/list/new_list = new_data
-		data = new_list.Copy(1,min( IC_MAX_LIST_LENGTH+1, new_list.len ))
+		data = new_list.Copy(max(1,new_list.len - IC_MAX_LIST_LENGTH+1),new_list.len +1)
```
Comment (ninjanomnom) [inline code/modules/integrated_electronics/core/pins.dm]: the second copy arg can just be 0 if you want the end of the list.

Is it intentional that you're cutting off the front of the list to conform with max list length?

File: code/modules/integrated_electronics/core/special_pins/list_pin.dm [hunk 2]
```diff
@@ -110,7 +110,7 @@
 /datum/integrated_io/lists/write_data_to_pin(var/new_data)
 	if(islist(new_data))
 		var/list/new_list = new_data
-		data = new_list.Copy(1,min( IC_MAX_LIST_LENGTH+1, new_list.len ))
+		data = new_list.Copy(max(1,new_list.len - IC_MAX_LIST_LENGTH+1),new_list.len +1)
```
Comment (ninjanomnom) [inline code/modules/integrated_electronics/core/special_pins/list_pin.dm]: same as above

File: code/modules/integrated_electronics/subtypes/input.dm [hunk 3]
```diff
@@ -794,13 +794,16 @@
...
-				set_pin_data(IC_OUTPUT, 2, cell.maxcharge)
-				set_pin_data(IC_OUTPUT, 3, cell.percent())
-	push_data()
+			if(get_turf(AM) in view(A))
+				var/ch=C.charge
+				var/mc=C.maxcharge
+				var/per=C.percent()
+				set_pin_data(IC_OUTPUT, 1, ch)
+				set_pin_data(IC_OUTPUT, 2, mc)
+				set_pin_data(IC_OUTPUT, 3, per)
```
Comment (ninjanomnom) [inline code/modules/integrated_electronics/subtypes/input.dm]: adding these new vars here is pointless

File: code/modules/integrated_electronics/subtypes/lists.dm [hunk 4]
```diff
@@ -138,9 +138,11 @@
 	var/list/input_list = get_pin_data(IC_INPUT, 1)
 	var/index = get_pin_data(IC_INPUT, 2)
 	var/item = get_pin_data(IC_INPUT, 3)
-	if(!islist(item))				//crh proof
-		input_list[index] = item
-		set_pin_data(IC_OUTPUT, 1, input_list)
+	var/list/red_list = list()
```
Comment (ninjanomnom) [inline code/modules/integrated_electronics/subtypes/lists.dm]: you don't need to create an empty list here if you're going to copy the other list

File: code/modules/integrated_electronics/subtypes/power.dm [hunk 5]
```diff
@@ -51,26 +51,29 @@
 			if(A.Adjacent(B))
 				if(AM.loc != assembly)
 					transfer_amount *= 0.8 // Losses due to distance.
-
+				var/ch=cell.charge
+				var/mc=cell.maxcharge
+				var/per=cell.percent()
+				set_pin_data(IC_OUTPUT, 1, ch)
+				set_pin_data(IC_OUTPUT, 2, mc)
+				set_pin_data(IC_OUTPUT, 3, per)
```
Comment (ninjanomnom) [inline code/modules/integrated_electronics/subtypes/power.dm]: don't make new vars like this. It just obfuscates code

**Review by ninjanomnom — CHANGES_REQUESTED** [review]

File: code/modules/integrated_electronics/core/pins.dm (hunk 1 above)
Comment (arsserpentarium) [reply code/modules/integrated_electronics/core/pins.dm]: yes,it's intentional for some reason.First option don't work.

File: code/modules/integrated_electronics/core/special_pins/list_pin.dm (hunk 2 above)
Comment (arsserpentarium) [reply code/modules/integrated_electronics/core/special_pins/list_pin.dm]: yes,it's intentional for some reason.First option don't work.

File: code/modules/integrated_electronics/subtypes/power.dm [hunk 6]
```diff
@@ -51,26 +51,26 @@
...
 		else
 			set_pin_data(IC_OUTPUT, 1, null)
 			set_pin_data(IC_OUTPUT, 2, null)
 			set_pin_data(IC_OUTPUT, 3, null)
 			activate_pin(2)
 			push_data()
 			return FALSE
-	return FALSE
+	else
+		return FALSE
```
Comment (ninjanomnom) [inline code/modules/integrated_electronics/subtypes/power.dm]: just do `if(!AM)` at the top so you don't have to indent this much and hurt readability of when this line happens

**Review by ninjanomnom — CHANGES_REQUESTED** [review]

**Review by ninjanomnom — APPROVED** [review]

**Pull request merged**